### PR TITLE
[Refactor] Refactor standard roi head with boxlist.

### PR DESCRIPTION
**Pull request by jbwang1997**
Thanks for your contribution and we appreciate it a lot. The following instructions would make your pull request more healthy and more easily get feedback. If you do not understand some items, don't worry, just make the pull request and seek help from maintainers.

## Motivation

Refactor standard roi head with boxlist
wait for merging of #8625

## Modification

Please briefly describe what modification is made in this PR.

## BC-breaking (Optional)

Does the modification introduce changes that break the backward-compatibility of the downstream repos?
If so, please describe how it breaks the compatibility and how the downstream projects should modify their code to keep compatibility with this PR.

## Use cases (Optional)

If this PR introduces a new feature, it is better to list some use cases here, and update the documentation.

## Checklist

1. Pre-commit or other linting tools are used to fix the potential lint issues.
2. The modification is covered by complete unit tests. If not, please add more unit test to ensure the correctness.
3. If the modification has potential influence on downstream projects, this PR should be tested with downstream projects, like MMDet or MMCls.
4. The documentation has been modified accordingly, like docstring or example tutorials.

### CODE REVIEW

File: mmdet/models/dense_heads/base_dense_head.py [hunk 1]
```diff
@@ -12,9 +12,10 @@
 from torch import Tensor
 
 from mmdet.structures import SampleList
+from mmdet.structures.bbox import BaseBoxes
 from mmdet.utils import InstanceList, OptMultiConfig
 from ..test_time_augs import merge_aug_results
-from ..utils import (filter_scores_and_topk, select_single_mlvl,
+from ..utils import (cat, filter_scores_and_topk, select_single_mlvl,
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/base_dense_head.py:18]: cat -> cat_boxes

File: mmdet/models/roi_heads/cascade_roi_head.py [hunk 2]
```diff
@@ -496,9 +496,11 @@ def _refine_roi(self, x: Tuple[Tensor], rois: Tensor,
                 for i in range(len(batch_img_metas)):
                     if rois[i].shape[0] > 0:
                         bbox_label = cls_scores[i][:, :-1].argmax(dim=1)
-                        refined_rois = bbox_head.regress_by_class(
-                            rois[i], bbox_label, bbox_preds[i],
+                        refined_bboxes = bbox_head.regress_by_class(
```
Comment (RangiLyu) [inline mmdet/models/roi_heads/cascade_roi_head.py:499]: add some comments

File: mmdet/models/utils/misc.py [hunk 3]
```diff
@@ -140,6 +142,7 @@ def empty_instances(
             results.
         mask_thr_binary (int, float): mask binarization threshold.
             Defaults to 0.
+        box_type
```
Comment (chhluo) [inline mmdet/models/utils/misc.py:145]: add description for box_type and with_boxlist

File: mmdet/models/utils/misc.py [hunk 4]
```diff
@@ -412,7 +422,7 @@ def images_to_levels(target, num_levels):
 
     [target_img0, target_img1] -> [target_level0, target_level1, ...]
     """
-    target = torch.stack(target, 0)
+    target = stack_boxes(target, 0)
```
Comment (chhluo) [inline mmdet/models/utils/misc.py:425]: `stack_boxes` is not suitable semantically when target is label not bbox.

File: mmdet/models/task_modules/samplers/sampling_result.py [hunk 5]
```diff
@@ -104,19 +106,20 @@ def __init__(self,
...
-                gt_bboxes = gt_bboxes.view(-1, 4)
-            self.pos_gt_bboxes = gt_bboxes[self.pos_assigned_gt_inds.long(), :]
+                gt_bboxes = gt_bboxes.view(-1, box_dim)
+            self.pos_gt_bboxes = gt_bboxes[self.pos_assigned_gt_inds.long()]
 
     @property
     def priors(self):
         """torch.Tensor: concatenated positive and negative priors"""
-        return torch.cat([self.pos_priors, self.neg_priors])
+        return cat_boxes([self.pos_priors, self.neg_priors])
```
Comment (chhluo) [inline mmdet/models/task_modules/samplers/sampling_result.py:122]: should update line 154: `if isinstance(value, torch.Tensor):`

File: mmdet/models/task_modules/coders/yolo_bbox_coder.py [hunk 6]
```diff
@@ -67,6 +70,7 @@ def decode(self, bboxes, pred_bboxes, stride):
         Returns:
             torch.Tensor: Decoded boxes.
         """
+        bboxes = get_box_tensor(bboxes)
```
Comment (chhluo) [inline mmdet/models/task_modules/coders/yolo_bbox_coder.py:73]: Update the typehint in docstring

File: mmdet/models/task_modules/assigners/iou2d_calculator.py [hunk 7]
```diff
@@ -39,6 +40,8 @@ def __call__(self, bboxes1, bboxes2, mode='iou', is_aligned=False):
         Returns:
             Tensor: shape (m, n) if ``is_aligned `` is False else shape (m,)
         """
+        bboxes1 = get_box_tensor(bboxes1)
```
Comment (chhluo) [inline mmdet/models/task_modules/assigners/iou2d_calculator.py:43]: Update typehint in docstring

File: mmdet/models/utils/misc.py (hunk 4 above)
Comment (jbwang1997) [reply mmdet/models/utils/misc.py:424]: I consider the name `stack_boxes` is uniform with other functions.

**Comment by zytx121** [comment]
LGTM

**Review by chhluo — DISMISSED** [review]

File: mmdet/structures/bbox/transforms.py [hunk 8]
```diff
@@ -88,12 +90,16 @@ def bbox2roi(bbox_list: List[Tensor]) -> Tensor:
         Tensor: shape (n, 5), [batch_ind, x1, y1, x2, y2]
     """
     rois_list = []
+    box_dim = 4
```
Comment (ZwwWayne) [inline mmdet/structures/bbox/transforms.py:93]: consider not using the default box_dim=4

File: mmdet/structures/bbox/transforms.py [hunk 9]
```diff
@@ -88,12 +90,16 @@ def bbox2roi(bbox_list: List[Tensor]) -> Tensor:
...
     for img_id, bboxes in enumerate(bbox_list):
+        if isinstance(bboxes, BaseBoxes):
+            box_dim = bboxes.box_dim
+            bboxes = bboxes.tensor
```
Comment (ZwwWayne) [inline mmdet/structures/bbox/transforms.py:97]:
```suggestion
        boxes = get_box_tensor(boxes)
        box_dim = boxes.size(-1)
```

Comment (ZwwWayne) [reply mmdet/structures/bbox/transforms.py:97]: We can also consider using 
```
boxes, box_dim = get_box_tensor(boxes, return_dim=True)
```

File: mmdet/structures/bbox/base_boxes.py [hunk 10]
```diff
@@ -91,6 +91,22 @@ def convert_to(self, dst_type: Union[str, type]) -> 'BaseBoxes':
         from .box_type import convert_box_type
         return convert_box_type(self, dst_type=dst_type)
 
+    def create_empty_box(self: T,
```
Comment (ZwwWayne) [inline mmdet/structures/bbox/base_boxes.py:94]: create_empty_box -> empty_boxes
also rename line in 110 from `create_fake_boxes` to `fake_boxes` to be consistent with Tensor.new_zeros().

File: mmdet/structures/bbox/transforms.py (hunk 8 above)
Comment (ZwwWayne) [reply mmdet/structures/bbox/transforms.py:93]: also update the docstring

File: mmdet/structures/bbox/transforms.py [hunk 11]
```diff
@@ -88,12 +90,16 @@ def bbox2roi(bbox_list: List[Tensor]) -> Tensor:
...
         if bboxes.size(0) > 0:
             img_inds = bboxes.new_full((bboxes.size(0), 1), img_id)
-            rois = torch.cat([img_inds, bboxes[:, :4]], dim=-1)
+            rois = torch.cat([img_inds, bboxes[:, :box_dim]], dim=-1)
```
Comment (ZwwWayne) [inline mmdet/structures/bbox/transforms.py:100]: 1. Is there any case that the boxes have more dimension?
2. considering line 102, I think the answer for Q1 is no, then we can simplify the code.

Comment (jbwang1997) [reply mmdet/structures/bbox/transforms.py:100]: I suppose `bboxes` must have `ndim == 2` and have no extra dimension. Then, the code can be simplified as below:

```
    for img_id, bboxes in enumerate(bbox_list):
        bboxes = get_box_tensor(bboxes)
        img_inds = bboxes.new_full((bboxes.size(0), 1), img_id)
        rois = torch.cat([img_inds, bboxes], dim=-1)
        rois_list.append(rois)
```

**Comment by jbwang1997** [comment]
I also move all box operations into mmdet.structures.bbox.

**Review by RangiLyu — COMMENTED** [review]
The word "boxlist" should be changed.

**Comment by jbwang1997** [comment]
> The word "boxlist" should be changed.

I have replaced all `boxlist` with `box type`.

**Review by RangiLyu — DISMISSED** [review]

File: mmdet/models/dense_heads/rpn_head.py [hunk 12]
```diff
@@ -291,7 +294,10 @@ def _bbox_post_process(self,
         else:
             # To avoid some potential error
             results_ = InstanceData()
-            results_.bboxes = results.scores.new_zeros(0, 4)
+            if isinstance(results.bboxes, BaseBoxes):
```
Comment (ZwwWayne) [inline mmdet/models/dense_heads/rpn_head.py:297]: can implement a function named `empty_box_as(results.bboxes)`

**Review by ZwwWayne — APPROVED** [review]

**Comment by zytx121** [comment]
LGTM

**Review by chhluo — COMMENTED** [review]
LGTM. Please rebase it on the latest dev-3.x

**Review by RangiLyu — APPROVED** [review]
LGTM

**Review by ZwwWayne — APPROVED** [review]

**Review by ZwwWayne — APPROVED** [review]
